dali/python/nvidia/dali/plugin/paddle.py

```diff
@@ -0,0 +1,445 @@
+# Copyright (c) 2017-2019, NVIDIA CORPORATION. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import ctypes
+import logging
+import math
+
+import numpy as np
+
+from nvidia.dali import types
+from nvidia.dali.backend import TensorListCPU, TensorGPU, TensorListGPU
+from paddle import fluid
+
+dtype_map = {
+    "=?": fluid.core.VarDesc.VarType.BOOL,
+    "=e": fluid.core.VarDesc.VarType.FP16,
+    "=f": fluid.core.VarDesc.VarType.FP32,
+    "=d": fluid.core.VarDesc.VarType.FP64,
+    "=B": fluid.core.VarDesc.VarType.UINT8,
+    "=b": fluid.core.VarDesc.VarType.INT8,
+    "=h": fluid.core.VarDesc.VarType.INT16,
+    "=i": fluid.core.VarDesc.VarType.INT32,
+    "=q": fluid.core.VarDesc.VarType.INT64,
+    "=l": fluid.core.VarDesc.VarType.INT64
+}
+
+
+def to_paddle_type(tensor):
+    r"""
+    Get paddle dtype for given tensor or tensor list
+
+    Args:
+        tensor: tensor or tensor list
+
+    Returns: fluid.core.VarDesc.VarType
+    """
+    if isinstance(tensor, (TensorListCPU, TensorListGPU)):
+        tensor = tensor.at(0)
+    dtype = tensor.dtype
+    if callable(dtype):
+        dtype = dtype()
+    else:
+        dtype = '=' + dtype.char
+    return dtype_map[dtype]
+
+
+def feed_ndarray(dali_tensor, ptr):
+    """
+    Copy contents of DALI tensor to Paddle's Tensor.
+
+    Parameters
+    ----------
+    `dali_tensor` : dali.backend.TensorCPU or dali.backend.TensorGPU
+                    Tensor from which to copy
+    `ptr` : LoDTensor data pointer
+            Destination of the copy
+    """
+    c_type_pointer = ctypes.c_void_p(ptr)
+    dali_tensor.copy_to_external(c_type_pointer)
+    return ptr
+
+
+def recursive_length(tensor, lod_level):
+    def _recurse(data, result, level):
```

Comment:
Can you provide more insight about idea here and purpose of this function?

Reply:
Paddle has the concept of `LoDTensor`, which models variable length sequences ( think ragged tensor in Tensorflow). This part basically constructs the offsets at each dimension, which is needed to construct the `LoDTensor`. For more details, please refer to this [guide](https://www.paddlepaddle.org.cn/documentation/docs/en/user_guides/howto/basic_concept/index_en.html)

Reply:
Thanks for the reference.

Reply:
My pleasure.